### ISSUE TITLE
AppArmor DENIES swtpm pid file access

### ISSUE THREAD

**Issue by mvollmer** [cockpit-project/bots#3845]
Ubuntu bug: https://bugs.launchpad.net/ubuntu/+source/libvirt/+bug/1989100

**Comment by mvollmer** [comment]
ubuntu-stable
Ooops, it happened again<hr>
```
# ----------------------------------------------------------------------
# testConfigureBeforeInstall (__main__.TestMachinesCreate)
error: Failed to start network default
error: Requested operation is not valid: network is already active
..+.+.....+...+...............+.+..+....+.....+......+....+.....+....+.........+..+..........+........+.+.....+.......+.....+..........+........+.........+...+...+......+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*........+.......+.........+.....+.+.....+...+.+......+.....+......+.......+......+.........+.................+..........+...............+...+...+...+.....+.+.....+....+.....+....+..+.+..+...+....+...+..+.............+...+............+.....+...+...+.+......+.....+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.+..+.......+.........+..+...+...................+..+....+..+..........+...+.....+......+.+..+.+...........+.+....................+..........+...+..............+.+.........+......+..+..........+............+.................+...+...+...+.+...+.................+.+............+..+..........+.........+..+...+.+........+.......+...+...+.........+........+......+.+......+..............+.+.................+............+.........+................+........+......+...............+.........+............+.+...+............+..+...+.............+........................+............+........+..................+.............+..+...+......+...+.+...........+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
......+.+..+......+...................+...+...+.....+......+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.......+........+......+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.................+.+........+......+.+....................+....+.....+.....................+.......+...+..............+.+.....+....+...........+.............+.....+......+.+...+..+.........+...+.............+........+....+...........+.+...............+......+........+....+........+..............................+....+...+...........+....+...+..+.......+.....+......+......................+..+.........+......+.+...+...............+..............+......+.+.....+...+............+...+...+...+.........+...+....+.........+..+.........+...+...+....+.........+.........+...+...+.........+...+...+.....+......+...+...+...+............+....+......+...+..+...+......+...+.+...............+......+...+...........+....+...+.........+......+..+.......+.....+.+.....+.......+..+.+..............+.+..+.+......+...........+...+.......+.....+....+...+..+..................+.............+..+...+.......+........+...+................+...+...........+............+...+.+........+.+........................+..+.+.........+.....................+..+...+....+...+........+...+............+....+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
-----
rehash: warning: skipping ca-certificates.crt,it does not contain exactly one certificate or CRL

DevTools listening on ws://127.0.0.1:10118/devtools/browser/fd7d622d-625b-4448-9af2-340667628385
ATTENTION: default value of option allow_rgb10_configs overridden by environment.
[0908/131347.142405:WARNING:bluez_dbus_manager.cc(247)] Floss manager not present, cannot set Floss enable/disable.
[0908/131347.187596:WARNING:sandbox_linux.cc(376)] InitializeSandbox() called with multiple threads in process gpu-process.
[0908/131347.659432:ERROR:devtools_http_handler.cc(636)] Using unsafe HTTP verb GET to invoke /json/new. This action will stop supporting GET and POST verbs in future versions.
> error: You are currently using minified code outside of NODE_ENV === "production". This means that you are running a slower development build of Redux. You can use loose-envify (https://github.com/zertosh/loose-envify) for browserify or setting mode to production in webpack (https://webpack.js.org/concepts/mode/) to ensure you have the correct code for your production build.
> error: Scrollbar test exception: TypeError: Cannot read properties of null (reading 'appendChild')
> log: osinfo-detect command failed:  (process:3665): GLib-GIO-WARNING **: 10:14:00.591: Can't find module 'gvfs' specified in GIO_USE_VFS
Traceback (most recent call last):
  File "<string>", line 21, in <module>
gi.repository.GLib.GError: osinfo-media-error: No volume descriptors (0)
error: failed to get domain 'VmNotInstalled'
error: failed to get domain 'VmNotInstalled'
> error: {"problem":null,"exit_status":1,"exit_signal":null,"message":"error: Failed to destroy domain 'VmNotInstalled'\nerror: Requested operation is not valid: domain is not running\nERROR    internal error: Could not get process id of swtpm\nDomain installation does not appear to have been successful.\nIf it was, you can restart your domain by running:\n  virsh --connect qemu:///system start VmNotInstalled\notherwise, please restart your installation.\nCommand '['virt-install', '--connect', 'qemu:///system', '--quiet', '--name', 'VmNotInstalled', '--os-variant', 'fedora28', '--memory', 'memory=130,maxmemory=150', '--check', 'path_in_use=off', '--wait', '-1', '--noautoconsole', '--disk', 'path=/var/lib/libvirt/images/VmNotInstalled.qcow2,device=disk,boot_order=2,bus=virtio,readonly=off,shareable=off,discard=unmap,driver_name=qemu,driver_type=qcow2', '--network', 'network=default,mac=52:54:00:70:46:83,model=virtio,boot_order=1,link_state=up', '--network', 'bridge=virbr0,mac=52:54:00:c2:24:37,model=rtl8139,link_state=up', '--graphics', 'type=vnc,listen=127.0.0.1,port=-1', '--graphics', 'type=spice,listen=127.0.0.1', '--cdrom', '/var/lib/libvirt/novell.iso', '--autostart', '--vcpus', 'vcpus=2,maxvcpus=8,sockets=2,cores=2,threads=2', '--boot', 'uefi']' returned non-zero exit status 1.\nTraceback (most recent call last):\n  File \"<stdin>\", line 304, in install_vm\n  File \"/usr/lib/python3.10/subprocess.py\", line 420, in check_output\n    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,\n  File \"/usr/lib/python3.10/subprocess.py\", line 524, in run\n    raise CalledProcessError(retcode, process.args,\nsubprocess.CalledProcessError: Command '['virt-install', '--connect', 'qemu:///system', '--quiet', '--name', 'VmNotInstalled', '--os-variant', 'fedora28', '--memory', 'memory=130,maxmemory=150', '--check', 'path_in_use=off', '--wait', '-1', '--noautoconsole', '--disk', 'path=/var/lib/libvirt/images/VmNotInstalled.qcow2,device=disk,boot_order=2,bus=virtio,readonly=off,shareable=off,discard=unmap,driver_name=qemu,driver_type=qcow2', '--network', 'network=default,mac=52:54:00:70:46:83,model=virtio,boot_order=1,link_state=up', '--network', 'bridge=virbr0,mac=52:54:00:c2:24:37,model=rtl8139,link_state=up', '--graphics', 'type=vnc,listen=127.0.0.1,port=-1', '--graphics', 'type=spice,listen=127.0.0.1', '--cdrom', '/var/lib/libvirt/novell.iso', '--autostart', '--vcpus', 'vcpus=2,maxvcpus=8,sockets=2,cores=2,threads=2', '--boot', 'uefi']' returned non-zero exit status 1.\nTraceback (most recent call last):\n  File \"<stdin>\", line 374, in <module>\n  File \"<stdin>\", line 315, in install_vm\n  File \"<stdin>\", line 304, in install_vm\n  File \"/usr/lib/python3.10/subprocess.py\", line 420, in check_output\n    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,\n  File \"/usr/lib/python3.10/subprocess.py\", line 524, in run\n    raise CalledProcessError(retcode, process.args,\nsubprocess.CalledProcessError: Command '['virt-install', '--connect', 'qemu:///system', '--quiet', '--name', 'VmNotInstalled', '--os-variant', 'fedora28', '--memory', 'memory=130,maxmemory=150', '--check', 'path_in_use=off', '--wait', '-1', '--noautoconsole', '--disk', 'path=/var/lib/libvirt/images/VmNotInstalled.qcow2,device=disk,boot_order=2,bus=virtio,readonly=off,shareable=off,discard=unmap,driver_name=qemu,driver_type=qcow2', '--network', 'network=default,mac=52:54:00:70:46:83,model=virtio,boot_order=1,link_state=up', '--network', 'bridge=virbr0,mac=52:54:00:c2:24:37,model=rtl8139,link_state=up', '--graphics', 'type=vnc,listen=127.0.0.1,port=-1', '--graphics', 'type=spice,listen=127.0.0.1', '--cdrom', '/var/lib/libvirt/novell.iso', '--autostart', '--vcpus', 'vcpus=2,maxvcpus=8,sockets=2,cores=2,threads=2', '--boot', 'uefi']' returned non-zero exit status 1."}
> warning: transport closed: disconnected
Traceback (most recent call last):
  File "/usr/lib64/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/home/mvo/work/cockpit-machines/test/check-machines-create", line 1865, in testConfigureBeforeInstall
    wait(lambda: "VmNotInstalled" in m.execute("virsh list --persistent"), delay=3)
  File "/home/mvo/work/cockpit-machines/test/common/testlib.py", line 2341, in wait
    raise Error(msg or "Condition did not become true.")
testlib.Error: Condition did not become true.
Wrote screenshot to TestMachinesCreate-testConfigureBeforeInstall-ubuntu-stable-127.0.0.2-2301-FAIL.png
Wrote HTML dump to TestMachinesCreate-testConfigureBeforeInstall-ubuntu-stable-127.0.0.2-2301-FAIL.html
Wrote JS log to TestMachinesCreate-testConfigureBeforeInstall-ubuntu-stable-127.0.0.2-2301-FAIL.js.log
Journal extracted to TestMachinesCreate-testConfigureBeforeInstall-ubuntu-stable-127.0.0.2-2301-FAIL.log.gz
audit: type=1400 audit(1662632058.096:33): apparmor="DENIED" operation="file_inherit" profile="swtpm" name="/run/libvirt/qemu/swtpm/1-VmNotInstalled-swtpm.pid" pid=3856 comm="swtpm" requested_mask="w" denied_mask="w" fsuid=118 ouid=0
Wrote JS log to TestMachinesCreate-testConfigureBeforeInstall-ubuntu-stable-127.0.0.2-2301-FAIL.js.log
Journal extracted to TestMachinesCreate-testConfigureBeforeInstall-ubuntu-stable-127.0.0.2-2301-FAIL.log.gz
umount: /usr/share/OVMF: not mounted.
mv: cannot stat '/var/lib/cockpittest/_etc_hosts': No such file or directory
umount: /usr/share/osinfo/os/fedoraproject.org/fedora-29.xml: not mounted.
umount: /usr/share/osinfo/os/fedoraproject.org/fedora-28.xml: not mounted.
bash: line 1: rpm: command not found
CDP: {"source":"rendering","level":"info","text":"Autofocus processing was blocked because a document already has a focused element.","timestamp":1662632245379.605,"url":"http://127.0.0.2:9191/machines#/vm?name=VmNotInstalled&connection=system"}
Warning: Stopping libvirtd.service, but it can still be activated by:
  libvirtd-admin.socket
  libvirtd-ro.socket
  libvirtd.socket
umount: /etc/libvirt: not mounted.
umount: /var/lib/libvirt: not mounted.
mv: cannot stat '/var/lib/cockpittest/_etc_crypttab': No such file or directory
mv: cannot stat '/var/lib/cockpittest/_etc_fstab': No such file or directory
mv: cannot stat '/var/lib/cockpittest/_etc_subgid': No such file or directory
mv: cannot stat '/var/lib/cockpittest/_etc_subuid': No such file or directory
mv: cannot stat '/var/lib/cockpittest/_etc_gshadow': No such file or directory
mv: cannot stat '/var/lib/cockpittest/_etc_shadow': No such file or directory
mv: cannot stat '/var/lib/cockpittest/_etc_group': No such file or directory
mv: cannot stat '/var/lib/cockpittest/_etc_passwd': No such file or directory
Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/test/check-machines-create", line 1865, in testConfigureBeforeInstall
    wait(lambda: "VmNotInstalled" in m.execute("virsh list --persistent"), delay=3)
  File "/home/mvo/work/cockpit-machines/test/common/testlib.py", line 2341, in wait
    raise Error(msg or "Condition did not become true.")
testlib.Error: Condition did not become true.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_hosts /etc/hosts']' returned non-zero exit status 1.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'umount -lf /etc/libvirt']' returned non-zero exit status 32.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'umount -lf /var/lib/libvirt']' returned non-zero exit status 32.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_crypttab /etc/crypttab']' returned non-zero exit status 1.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_fstab /etc/fstab']' returned non-zero exit status 1.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_subgid /etc/subgid']' returned non-zero exit status 1.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_subuid /etc/subuid']' returned non-zero exit status 1.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_gshadow /etc/gshadow']' returned non-zero exit status 1.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_shadow /etc/shadow']' returned non-zero exit status 1.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_group /etc/group']' returned non-zero exit status 1.

Traceback (most recent call last):
  File "/home/mvo/work/cockpit-machines/bots/machine/machine_core/ssh_connection.py", line 307, in execute
    res = subprocess.run(command_line,
  File "/usr/lib64/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['env', '-u', 'LANGUAGE', 'LC_ALL=C', 'ssh', '-p', '2301', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'IdentitiesOnly=yes', '-o', 'BatchMode=yes', '-o', 'PKCS11Provider=none', '-o', 'LogLevel=ERROR', '-l', 'root', '127.0.0.2', '-o', 'ControlPath=/tmp/.cockpit-test-resources/ssh-%h-%p-%r-1237531', 'set -e;', 'mv /var/lib/cockpittest/_etc_passwd /etc/passwd']' returned non-zero exit status 1.

# Result testConfigureBeforeInstall (__main__.TestMachinesCreate) failed
# 1 TEST FAILED [245s on thinner]
not ok 1 /home/mvo/work/cockpit-machines/test/check-machines-create TestMachinesCreate.testConfigureBeforeInstall [ND@1]
```
First occurrence: 2022-09-08T13:17:28.196889 | revision 968ecba7f8bf1a12acb1cae142854248a7f37399
Times recorded: 1
Latest occurrences:

- 2022-09-08T13:17:28.196889 | revision 968ecba7f8bf1a12acb1cae142854248a7f37399

**Comment by martinpitt** [comment]
Hmm, the [latest build](https://cockpit-logs.us-east-1.linodeobjects.com/image-refresh-logs/ubuntu-stable-20221205-225656.log) pulled in the released version 0.6.3-0ubuntu4 , but the fix is only in [-0ubuntu4.1](https://launchpad.net/ubuntu/+source/swtpm/0.6.3-0ubuntu4.1). That is still sitting in updates-proposed.

So apparently our tests changed to just not trigger this problem any more. :man_shrugging:
